pkg/querier/tripperware/instantquery/instant_query.go

```diff
@@ -0,0 +1,442 @@
+}
+
+type instantQueryCodec struct {
+	tripperware.Codec
+}
+
+func (resp *PrometheusInstantQueryResponse) HTTPHeaders() map[string][]string {
+	if resp != nil && resp.GetHeaders() != nil {
+		r := map[string][]string{}
+		for _, header := range resp.GetHeaders() {
+			if header != nil {
+				r[header.Name] = header.Values
+			}
+		}
+
+		return r
+	}
+	return nil
+}
+
+func (instantQueryCodec) DecodeRequest(_ context.Context, r *http.Request, forwardHeaders []string) (tripperware.Request, error) {
+	result := PrometheusRequest{Headers: map[string][]string{}}
+	var err error
+	result.Time, err = util.ParseTime(r.FormValue("time"))
+	if err != nil {
+		return nil, decorateWithParamName(err, "time")
+	}
+
+	result.Query = r.FormValue("query")
+	result.Stats = r.FormValue("stats")
+	result.Path = r.URL.Path
+
+	// Include the specified headers from http request in prometheusRequest.
+	for _, header := range forwardHeaders {
+		for h, hv := range r.Header {
+			if strings.EqualFold(h, header) {
+				result.Headers[h] = hv
+				break
+			}
+		}
+	}
+
+	return &result, nil
+}
+
+func (instantQueryCodec) DecodeResponse(ctx context.Context, r *http.Response, _ tripperware.Request) (tripperware.Response, error) {
+	if r.StatusCode/100 != 2 {
+		body, _ := io.ReadAll(r.Body)
+		return nil, httpgrpc.Errorf(r.StatusCode, string(body))
+	}
+
+	log, ctx := spanlogger.New(ctx, "PrometheusInstantQueryResponse") //nolint:ineffassign,staticcheck
+	defer log.Finish()
+
+	buf, err := tripperware.BodyBuffer(r)
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+	var resp PrometheusInstantQueryResponse
+	if err := json.Unmarshal(buf, &resp); err != nil {
+		return nil, httpgrpc.Errorf(http.StatusInternalServerError, "error decoding response: %v", err)
+	}
+
+	for h, hv := range r.Header {
+		resp.Headers = append(resp.Headers, &tripperware.PrometheusResponseHeader{Name: h, Values: hv})
+	}
+	return &resp, nil
+}
+
+func (instantQueryCodec) EncodeRequest(ctx context.Context, r tripperware.Request) (*http.Request, error) {
+	promReq, ok := r.(*PrometheusRequest)
+	if !ok {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, "invalid request format")
+	}
+	params := url.Values{
+		"time":  []string{tripperware.EncodeTime(promReq.Time)},
+		"query": []string{promReq.Query},
+	}
+
+	if promReq.Stats != "" {
+		params.Add("stats", promReq.Stats)
+	}
+
+	u := &url.URL{
+		Path:     promReq.Path,
+		RawQuery: params.Encode(),
+	}
+
+	var h = http.Header{}
+
+	for n, hv := range promReq.Headers {
+		for _, v := range hv {
+			h.Add(n, v)
+		}
+	}
+
+	req := &http.Request{
+		Method:     "GET",
+		RequestURI: u.String(), // This is what the httpgrpc code looks at.
+		URL:        u,
+		Body:       http.NoBody,
+		Header:     h,
+	}
+
+	return req.WithContext(ctx), nil
+}
+
+func (instantQueryCodec) EncodeResponse(ctx context.Context, res tripperware.Response) (*http.Response, error) {
+	sp, _ := opentracing.StartSpanFromContext(ctx, "APIResponse.ToHTTPResponse")
```

Comment:
`APIResponse` sounds vague. Shall we make it more clear like instant query APIResponse

Reply:
hummm... maybe... we know that this is an instant query call right? 

So the question is.. Should we keep the same span we have on the query_range (so will know what to look on both cases) or should we have different span names for each one of them?

Reply:
I don't know if we have any use cases like aggregating spans by their name. In this case probably separating the name would be better. If we are just looking at individual traces yeah then let's keep the same name.